modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java

```diff
@@ -273,6 +275,103 @@ public void testIndexCreationWhenNodeStopped() throws Exception {
+    public void testIndexCreationWhenNodeIsStoppedWithoutCheckpoint() throws Exception {
+        // Start topology.
+        Ignite ignite = startGrids(2);
+
+        ignite.cluster().active(true);
+
+        // Create table, add some data.
+        int entryCnt = 50;
+
+        try (Connection conn = DriverManager.getConnection("jdbc:ignite:thin://127.0.0.1:10800")) {
+            executeJdbc(conn,
+                " CREATE TABLE PERSON (\n" +
+                    " FIRST_NAME VARCHAR,\n" +
+                    " LAST_NAME VARCHAR,\n" +
+                    " ADDRESS VARCHAR,\n" +
+                    " LANG VARCHAR,\n" +
+                    " BIRTH_DATE TIMESTAMP,\n" +
+                    " CONSTRAINT PK_PERSON PRIMARY KEY (FIRST_NAME,LAST_NAME,ADDRESS,LANG)\n" +
+                    " ) WITH \"key_type=PersonKeyType, " +
+                    "CACHE_NAME=PersonCache, value_type=PersonValueType, AFFINITY_KEY=FIRST_NAME,template=PARTITIONED,backups=1\"");
+
+            try (PreparedStatement stmt = conn.prepareStatement(
+                "insert into Person(LANG, FIRST_NAME, ADDRESS, LAST_NAME, BIRTH_DATE) values(?,?,?,?,?)")) {
+                for (int i = 0; i < entryCnt; i++) {
+                    String s = String.valueOf(i);
+
+                    stmt.setString(1, s);
+                    stmt.setString(2, s);
+                    stmt.setString(3, s);
+                    stmt.setString(4, s);
+                    stmt.setTimestamp(5, new Timestamp(System.currentTimeMillis()));
+
+                    stmt.executeUpdate();
+                }
+            }
+        }
+
+        // Stop second node.
+        stopGrid(1, true);
+
+        // Create an index on remaining node.
+        try (Connection conn = DriverManager.getConnection("jdbc:ignite:thin://127.0.0.1:10800")) {
+            executeJdbc(conn, "create index PERSON_FIRST_NAME_IDX on PERSON(FIRST_NAME)");
+        }
+
+        // Restart second node.
+        startGrid(1);
+
+        // Await for index rebuild on started node.
+        assert GridTestUtils.waitForCondition(() -> {
+            try (Connection conn = DriverManager.getConnection("jdbc:ignite:thin://127.0.0.1:10801")) {
+                try (PreparedStatement stmt = conn.prepareStatement(
+                    "EXPLAIN SELECT * FROM Person USE INDEX(PERSON_FIRST_NAME_IDX) WHERE FIRST_NAME=?")) {
+                    stmt.setString(1, String.valueOf(1));
+
+                    StringBuilder fullPlan = new StringBuilder();
+
+                    try (ResultSet rs = stmt.executeQuery()) {
+                        while (rs.next())
+                            fullPlan.append(rs.getString(1)).append("; ");
+                    }
+
+                    System.out.println("PLAN: " + fullPlan);
+
+                    return fullPlan.toString().contains("PUBLIC.PERSON_FIRST_NAME_IDX");
+                }
+            }
+            catch (Exception e) {
+                throw new RuntimeException("Query failed.", e);
```

Comment:
I'd rather have a `fail(...)` here

Reply:
we will lose the original exception then